### PR TITLE
Added model training and evaluation functionality

### PR DESCRIPTION
**Pull request by kmzzhang**
Added new class train() in deepCR.training for training new model
Example usage:
from deepCR import train
trainer = train(image, mask, epoch=5, device='GPU'), where image and mask are 3D numpy arrays of shape (N,W,W)

For automatic evaluation:
from deepCR import roc
tpr, fpr = deepCR.roc(model, image_test, mask_test) where model is a deepCR instance and image_test and mask_test are 3D numpy arrays of shape (N,W,W) from test set.

### CODE REVIEW

**Comment by profjsb** [comment]
See inline comments. Also, should be easy to add test coverage of `dataset.py`.

File: deepCR/__init__.py [hunk 1]
```diff
@@ -10,7 +10,9 @@
 
 """
 from deepCR.model import deepCR
+from deepCR.training import train
+from deepCR.evaluate import roc
 
-__all__ = model.__all__
+__all__ = [deepCR, train, roc]
 
 __version__ = '0.1.4'
```
Comment (profjsb) [inline deepCR/__init__.py]: push the version number to 0.1.5

File: deepCR/dataset.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+from torch.utils.data import Dataset
+import numpy as np
```
Comment (profjsb) [inline deepCR/dataset.py]: put `numpy` and standard imports before `pytorch` (as per PEP8)

File: deepCR/dataset.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+class dataset(Dataset):
+    def __init__(self, image, mask, ignore=None, sky=None, part='train', aug_sky=[0, 0]):
+        """ custom pytorch dataset class to load deepCR-mask training data
+        :param image: image with CR
+        :param mask: CR mask
+        :param ignore: loss mask, e.g., bad pixel, saturation, etc.
+        :param sky: (np.ndarray) [N,] sky background level
+        :param part: either 'train' or 'val'. split by 0.8, 0.2
+        :param aug_sky: [negative number, positive number]. Add sky background by aug_sky[0] * sky to aug_sky[1] * sky.
```
Comment (profjsb) [inline deepCR/dataset.py]: better description about what this parameter is and why it is needed.

File: deepCR/dataset.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import numpy as np
+
+class dataset(Dataset):
+    def __init__(self, image, mask, ignore=None, sky=None, part='train', aug_sky=[0, 0]):
+        """ custom pytorch dataset class to load deepCR-mask training data
+        :param image: image with CR
+        :param mask: CR mask
+        :param ignore: loss mask, e.g., bad pixel, saturation, etc.
+        :param sky: (np.ndarray) [N,] sky background level
+        :param part: either 'train' or 'val'. split by 0.8, 0.2
```
Comment (profjsb) [inline deepCR/dataset.py]: is 0.8 and 0.2 hard coded? Shouldn't this be up to the user?

File: deepCR/dataset.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        :param aug_sky: [negative number, positive number]. Add sky background by aug_sky[0] * sky to aug_sky[1] * sky.
+        """
+        np.random.seed(1)
+        len = image.shape[0]
+        if sky is None:
+            sky = np.zeros_like(image)
+        if ignore is None:
+            ignore = np.zeros_like(image)
+        if part == 'train':
+            self.image = image[:int(len * 0.8)]
```
Comment (profjsb) [inline deepCR/dataset.py]: use the `numpy.s_` operator for repeated slices.

File: deepCR/evaluate.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+import numpy as np
+from deepCR.util import maskMetric
```
Comment (profjsb) [inline deepCR/evaluate.py]: local imports go last in the import order

File: deepCR/evaluate.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    for t in tqdm(range(len(data))):
+        dat = data[t]
+        pdt_mask = model.clean(dat[0], inpaint=False, binary=False)
+        msk = dat[1]
+        ignore = dat[2]
+        for i in range(nROC):
+            binary_mask = np.array(pdt_mask > thresholds[i]) * (1 - ignore)
+            metric[i] += maskMetric(binary_mask, msk * (1 - ignore))
+    TP, TN, FP, FN = metric[:, 0], metric[:, 1], metric[:, 2], metric[:, 3]
+    tpr = TP / (TP + FN)
```
Comment (profjsb) [inline deepCR/evaluate.py]: this will error if TP + FN = 0. Catch it?

File: deepCR/evaluate.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+__all__ = 'roc'
+
+
+def _roc(model, data, thresholds):
+    """ internal function called by roc
+
+    :param model:
+    :param data: deepCR.dataset object
+    :param thresholds:
+    :return: tpr, fpr
```
Comment (profjsb) [inline deepCR/evaluate.py]: note that these parameters are returned in percentages not decimals (since you multiply them by 100).

File: deepCR/model.py [hunk 9]
```diff
@@ -212,7 +205,7 @@ def clean_large_parallel(self, img0, threshold=0.5, inpaint=True, binary=True,
         mask = np.memmap(mask_filename_memmap, dtype=np.int8 if binary else img0_dtype,
                            shape=im_shape, mode='w+')
 
-        @wrap_non_picklable_objects
+        #@wrap_non_picklable_objects
```
Comment (profjsb) [inline deepCR/model.py]: how does parallel stitching work without this decorator?!

File: deepCR/training.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+            self.dint = torch.cuda.ByteTensor
+            self.network = nn.DataParallel(UNet2Sigmoid(1,1,hidden))
+            self.network.type(self.dtype)
+        else:
+            self.dtype = torch.FloatTensor
+            self.dint = torch.ByteTensor
+            self.network = WrappedModel(UNet2Sigmoid(1,1,hidden))
+            self.network.type(self.dtype)
+
+        self.optimizer = optim.Adam(self.network.parameters(), lr=lr)
```
Comment (profjsb) [inline deepCR/training.py]: consider making `optimizer` and `lr_scheduler` as keywords supplied  by the user, with sensible defaults.

File: deepCR/training.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+            self.set_input(*dat)
+            self.pdt_mask = self.network(self.img0)
+            loss = self.backward_network()
+            lmask += float(loss.detach()) * n
+            metric += maskMetric(self.pdt_mask.reshape(-1, self.shape, self.shape).detach().cpu().numpy() > 0.5, dat[1].numpy())
+        lmask /= count
+        TP, TN, FP, FN = metric[0], metric[1], metric[2], metric[3]
+        TPR = TP / (TP + FN)
+        FPR = FP / (FP + TN)
+        print('[TPR=%.3f, FPR=%.3f] @threshold = 0.5' % (TPR * 100, FPR * 100))
```
Comment (profjsb) [inline deepCR/training.py]: perhaps only do print statements if `verbose` = True

File: deepCR/util.py [hunk 12]
```diff
@@ -24,3 +24,17 @@ def medmask(image, mask):
         else:
             clean[x, y] = medianImage
     return clean
+
+def maskMetric(PD, GT):
+    if len(PD.shape) == 2:
```
Comment (profjsb) [inline deepCR/util.py]: raise exceptions if the shapes are not what you are expecting.

File: deepCR/training.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+from deepCR.util import maskMetric
+from deepCR.dataset import dataset
+
+__all__ = 'train'
+
+class train():
+
+    def __init__(self, image, mask, ignore=None, sky=None, name='model', hidden=32, gpu=False, epoch=50, batch_size=16, lr=0.005, aug_sky=[0, 0], save_after=0, plot_every=10, directory='./'):
+        """ This is the class for training deepCR-mask.
+        :param image: np.ndarray (N*W*W) training data: image array with CR.
```
Comment (profjsb) [inline deepCR/training.py]: you might consider making this either an (N*W*W)  numpy array OR a pytorch Dataset (with a load from disk capability) -- a preloaded numpy array will limit the user to their available RAM.

File: deepCR/util.py (hunk 12 above)
Comment (kmzzhang) [reply deepCR/util.py]: Is this necessary for internal functions (plus unit tests)?

**Comment by kmzzhang** [comment]
Made edits according to the comments.

File: deepCR/dataset.py (hunk 3 above)
Comment (profjsb) [reply deepCR/dataset.py]: Still not clear why this is needed.

File: deepCR/dataset.py (hunk 5 above)
Comment (profjsb) [reply deepCR/dataset.py]: The point was to reuse the slice, so
```python
self.image = image[np.s_[:int(len * f_train)]]
self.mask = mask[np.s_[:int(len * f_train)]]
...
```
becomes:
```python
the_slice = np.s_[:int(len * f_train)]
self.image = image[the_slice]
self.mask = mask[the_slice]
....
```

**Comment by profjsb** [comment]
Accepting changes. Will open issues for outstanding items.
